src/librustc/hir/mod.rs

```diff
@@ -1486,6 +1488,10 @@ pub enum ExprKind {
     Cast(P<Expr>, P<Ty>),
     /// A type reference (e.g., `Foo`).
     Type(P<Expr>, P<Ty>),
+    /// Semantically equivalent to `{ let _t = expr; _t }`.
+    /// Maps directly to `hair::ExprKind::Use`.
+    /// Only exists to tweak the drop order in HIR.
```

Comment:
Can we have more docs on where this is introduced and why it exists? As it stands it's pretty confusing

Reply:
It is used in the `for` loop desugaring as seen in this PR and also in https://github.com/rust-lang/rust/pull/59288.

Reply:
I'm not asking you to explain it to me, I figured it out, I'm saying that this should be in the docs.

Reply:
Sure why not. Some parts haven't been merged yet so it would be slightly weird to talk about future events in the docs.

Reply:
I don't find the name... useful (sorry). I'd call this `SomethingScope`, which for the lack of a better option would be `TerminatingScope`.

I believe that `ExprKind::Use` elsewhere has nothing to do with scopes (just like `mir::Rvalue::Use` doesn't), so using `Use` is misleading.

Reply:
@eddyb Fwiw, I contemplated `ExprKind::Terminating`. I agree completely that `::Use` sucks but I didn't have anything super better at the time.